### ISSUE TITLE
CLI Script to understand PR Workflows

### ISSUE THREAD

**Issue by HarshCasper** [HarshCasper/Rotten-Scripts#756]
# Description

Create a CLI Script which can help us better understand the Pull Request Workflows on the Project with some important statistics: 

- How many PRs are open? 
- How long have the PRs been open? 
- Number of Comments on open PRs? 
- Merged and Closed PRs

You can come up with your own customization on what you are aiming to build. Some good resources to check out on how to build this can be found on the [Pull Request API](https://developer.github.com/v3/pulls/) by Github.

## Language

- [X] Python
- [X] Golang
- [X] Javascript

## Checklist:

- **Details on the Script**

| name            | about                                                  | labels  | assignees                                                      |
| --------------- | ------------------------------------------------------ | ------- | -------------------------------------------------------------- |
| PR Workflow | CLI to fetch PR information and their status quo | MWOC, hard, Python, Javascript, Golang |  |

**Comment by vismitap** [comment]
Hey! I would like to work on this issue as a part of DWoC. Please assign this to me!

**Comment by vybhav72954** [comment]
@vismitap Any Updates?

**Comment by vismitap** [comment]
Yeah sorry! Give me 2 days, I will work on it as a part of GSSOC.😅

**Comment by vybhav72954** [comment]
> Yeah sorry! Give me 2 days, I will work on it as a part of GSSOC.😅

Waiting :)

**Comment by rahulraikwar00** [comment]
Hey! I would like to work on this issue as a part of GSSOC21. Please assign this to me!

**Comment by vybhav72954** [comment]
@vismitap Can you confirm the status. @rahulraikwar00 I will assign if i will receive n update

**Comment by rahulraikwar00** [comment]
update ? what kind of?

**Comment by vybhav72954** [comment]
> update ? what kind of?

No update from @vismitap as the issue is already assigned to him.

**Comment by vismitap** [comment]
@vybhav72954 I am having my exams now, so could not complete. It's ok, please assign @rahulraikwar00. I will claim another issue in some days. So sorry!😣

**Comment by vybhav72954** [comment]
> @vybhav72954 I am having my exams now, so could not complete. It's ok, please assign @rahulraikwar00. I will claim another issue in some days. So sorry!😣

Nevermind, please focus on your exams for now, you are always welcome here. Do have a look at the open issues once your exams are over., and thank you for the timely update.

@rahulraikwar00  Assigning this to you. Best of Luck, in case of any problem feel free to ping me.

**Comment by rahulraikwar00** [comment]
What should I name this tool?
Like..
Example: prreport , prview, or prworkflow
Suggest me a memorable short name?

**Comment by vybhav72954** [comment]
Lets keep it prworkflow only XD

**Comment by rahulraikwar00** [comment]
@vybhav72954 there is 2 ways to get details about the project
1. Through full url : we can take full url of project from user .
2. Through username and project repo : ex :- "username/repo"

What should I do?

**Comment by vybhav72954** [comment]
> @vybhav72954 there is 2 ways to get details about the project
> 1. Through full url : we can take full url of project from user .
> 2. Through username and project repo : ex :- "username/repo"
> 
> What should I do? 

Username/Repo is a better option

**Comment by rahulraikwar00** [comment]
@vybhav72954  it Looks like this for now yet to add more options and functionality,
and one thing to ask is , can i cache the api response in a json file because it would make this really fast. 
![Screenshot_14](https://user-images.githubusercontent.com/54519734/110941773-dcd45700-835e-11eb-9ea3-226cafab607d.png)

**Comment by vybhav72954** [comment]
@HarshCasper @sohamsshah

**Comment by vybhav72954** [comment]
> @vybhav72954  it Looks like this for now yet to add more options and functionality,
> and one thing to ask is , can i cache the api response in a json file because it would make this really fast. 
> 

This looks good to me!!!

I think making a Json file is a good idea for API response. This will make the script more useful.

**Comment by rahulraikwar00** [comment]
@vybhav72954  i've completed the basic structure of tool prworkflow
should i push it ? i need to confirm whether it is going in the right direction ?

**Comment by HarshCasper** [comment]
@rahulraikwar00 In my opinion, you can make a PR and keep it as a draft. It would ensure that your progress is completely logged up on our side. Good work mate! 

Coming to the caching part, JSON is a good way.

**Comment by rahulraikwar00** [comment]
I don't know how to convert pull request into draft
And do I need to download whole project if I want to push my files?
I mean do I need to setup whole rotten-scripts project locally to work on it?

**Comment by HarshCasper** [comment]
> I don't know how to convert pull request into draft

I will do it for you. 

> I mean do I need to setup whole rotten-scripts project locally to work on it?

You just need to clone it up. No installations or anything required.

**Comment by vybhav72954** [comment]
Check the [CONTRIBUTING GUIDE](https://github.com/HarshCasper/Rotten-Scripts/blob/master/CONTRIBUTING.md). This will help you with setting up Rotten-Scripts.

talking about draft, I will do that, or you can do that yours as well!!! The option is on top right after you will make the PR.

**Comment by rahulraikwar00** [comment]
> > I don't know how to convert pull request into draft
> 
> I will do it for you. 
> 
> > I mean do I need to setup whole rotten-scripts project locally to work on it?
> 
> You just need to clone it up. No installations or anything required.
> 

Ya ..that is what I'm talking about is there any way to clone a perticular folder? Or have to clone whole project?

**Comment by HarshCasper** [comment]
@rahulraikwar00 : Whole Project

**Comment by vybhav72954** [comment]
Why you closed the PR?

**Comment by rahulraikwar00** [comment]
I've uploaded file directly 😅, some files were missing 
I'll do it  by tomorrow evening 👍🏻
And is it ok if i push files with virtualenv folder?

**Comment by vybhav72954** [comment]
I don't think we need virtualenv folder. Any specific reason you want to push it?

The `bin`, `lib` and all other directories shouldn't be pushed to repo, and you can keep them in your directory though `.gitginore` will take care of it.

**Comment by rahulraikwar00** [comment]
Some checks are failing! What should I do?
Spelling check! 😅 Prworkflow Don't make sense

**Comment by rahulraikwar00** [comment]
@vybhav72954 please review my pr

**Comment by vybhav72954** [comment]
@rahulraikwar00 Let me check

**Comment by rahulraikwar00** [comment]
@vybhav72954  These features has been added! Apart from these, if you want to add anything, please let me know
1. How many PRs are open? How long have the PRs been open?
2. Number of Comments on open PRs?
3. Merged PRs
4. Closed PRs

**Comment by vybhav72954** [comment]
@HarshCasper @sohamsshah @seema1711 WHat do you guys think?

**Comment by rahulraikwar00** [comment]
@vybhav72954 
@HarshCasper 
Can we access the "repository pulse" through GitHub api? if not is there any alternative?

**Comment by sohamsshah** [comment]
> @vybhav72954 These features has been added! Apart from these, if you want to add anything, please let me know
> 
> 1. How many PRs are open? How long have the PRs been open?
> 2. Number of Comments on open PRs?
> 3. Merged PRs
> 4. Closed PRs

Great work @rahulraikwar00 
I guess the features you have added are great. Curious, if we can access the tags/label on the PR, so more custom search/filter can be made on each feature? For example, Number of PRs open with tag "X" and Number of comments on open PR with tag "X" and so on. This can be a great use-case for a Repository for checking on PRs that are under "GSSoC 21" or "Under-review" or "Work-in-progress" or similar.

In a nutshell, the features are good to go, if this can be added its definitely cherry on the top!

**Comment by rahulraikwar00** [comment]
> > @vybhav72954 These features has been added! Apart from these, if you want to add anything, please let me know
> > 
> > 1. How many PRs are open? How long have the PRs been open?
> > 2. Number of Comments on open PRs?
> > 3. Merged PRs
> > 4. Closed PRs
> 
> Great work @rahulraikwar00 
> I guess the features you have added are great. Curious, if we can access the tags/label on the PR, so more custom search/filter can be made on each feature? For example, Number of PRs open with tag "X" and Number of comments on open PR with tag "X" and so on. This can be a great use-case for a Repository for checking on PRs that are under "GSSoC 21" or "Under-review" or "Work-in-progress" or similar.
> 
> In a nutshell, the features are good to go, if this can be added its definitely cherry on the top!
> 
>  

I was trying for almost 2 days..
But we can't access tags directly from pull request through GitHub api
We have to make request for each request to get labels
I'm trying some other things 
Like GitHub graphQL api
May be this functionality can be achieve by this ..
Give me couple days more.. I'll definitely come with a solution 👍🏻

**Comment by vybhav72954** [comment]
> > > @vybhav72954 These features has been added! Apart from these, if you want to add anything, please let me know
> > > 
> > > 1. How many PRs are open? How long have the PRs been open?
> > > 2. Number of Comments on open PRs?
> > > 3. Merged PRs
> > > 4. Closed PRs
> > 
> > 
> > Great work @rahulraikwar00
> > I guess the features you have added are great. Curious, if we can access the tags/label on the PR, so more custom search/filter can be made on each feature? For example, Number of PRs open with tag "X" and Number of comments on open PR with tag "X" and so on. This can be a great use-case for a Repository for checking on PRs that are under "GSSoC 21" or "Under-review" or "Work-in-progress" or similar.
> > In a nutshell, the features are good to go, if this can be added its definitely cherry on the top!
> 
> I was trying for almost 2 days..
> But we can't access tags directly from pull request through GitHub api
> We have to make request for each request to get labels
> I'm trying some other things
> Like GitHub graphQL api
> May be this functionality can be achieve by this ..
> Give me couple days more.. I'll definitely come with a solution 👍🏻

Please take your time

**Comment by rahulraikwar00** [comment]
@HarshCasper @vybhav72954 
GitHub graphQL api use authentication token instead of username and password 
Which is kinda make things complex 
To generate a token and thn login into script and thn use it.. 
I've tried to utf-8 base64 string for authentication header but it didn't worked
And a good news is that with data mining using graphQL we can get whatever we need in our script
Like ; tag , assignees, comments, counts, status, branch,...lot more.
Should I continue with token verification and later we will change the way the token is verified with the username and password when we find out? Or it you know how to do this please help me!

**Comment by rahulraikwar00** [comment]
@seema1711 can you please help me?

**Comment by HarshCasper** [comment]
@rahulraikwar00 To be simple, we can go with your existing script now. Next we can build on it to add more features to the same. What do you think? 

I'm personally fine with either of them.

**Comment by rahulraikwar00** [comment]
> @rahulraikwar00 To be simple, we can go with your existing script now. Next we can build on it to add more features to the same. What do you think? 
> 
> I'm personally fine with either of them.

But existing script does nothing except display open pulls .!
And now I have come forward very much, it is 10 times better than the one I was working on.
And it will fish by tomorrow evening, then we'll decide ..I think so.!
@HarshCasper  now it's a full fledged cli tool 
And the key features are:
  1. Can display - Open , close , merge
  2. Search with tags as suggested by @sohamsshah 
  3. It has verbose mode. By which it can display nested info like assignee, comments, review etc
4. Upper limit of API request is now 5000 request per hour which was previously 60.
What do you think?

**Comment by sohamsshah** [comment]
Good work 💪 @rahulraikwar00 I loved the never-give-up attitude haha. 

I agree to @HarshCasper on the point that if the implementation is getting more complicated and tedious, you can surely improve later; and current work is also good.

But I believe if you can implement such good feature, I encourage this and also more than happy to help, if you need any.

**Comment by rahulraikwar00** [comment]
> Good work 💪 @rahulraikwar00 I loved the never-give-up attitude haha. 
> 
> I agree to @HarshCasper on the point that if the implementation is getting more complicated and tedious, you can surely improve later; and current work is also good.
> 
> But I believe if you can implement such good feature, I encourage this and also more than happy to help, if you need any. 

Thank you 😁 
I don't worry about if this count as a single pull request.
The thing is during this project I learnt a lot things.
Sometimes it took me to reach up to 3rd page of Google 😂
So I don't want to let my "mehnat" waste.
So I want do this.
Yes I need help,
I searched a lot but unable to find the way to verify graphQL api using username and password.
Please let me know if you find.

**Comment by rahulraikwar00** [comment]
@sohamsshah can you please review it
